evidence/reactor_test.go

```diff
@@ -31,129 +33,176 @@ var (
+				require.False(t, r.IsRunning())
+			}
+		}
+
+		leaktest.Check(t)
 	})
 
 	return rts
 }
 
-func createTestSuites(t *testing.T, stateStores []sm.Store, chBuf uint) []*reactorTestSuite {
-	t.Helper()
+func (rts *reactorTestSuite) start(t *testing.T) {
+	rts.network.Start(t)
+	require.Len(t, rts.network.RandomNode().PeerManager.Peers(), rts.numStateStores-1,
+		"network does not have expected number of nodes")
+}
 
-	numSStores := len(stateStores)
-	testSuites := make([]*reactorTestSuite, numSStores)
-	evidenceTime := time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)
+func (rts *reactorTestSuite) waitForEvidence(t *testing.T, evList types.EvidenceList, ids ...p2p.NodeID) {
+	t.Helper()
 
-	for i := 0; i < numSStores; i++ {
-		logger := log.TestingLogger().With("validator", i)
-		evidenceDB := dbm.NewMemDB()
-		blockStore := &mocks.BlockStore{}
-		blockStore.On("LoadBlockMeta", mock.AnythingOfType("int64")).Return(
-			&types.BlockMeta{Header: types.Header{Time: evidenceTime}},
+	fn := func(pool *evidence.Pool) {
+		var (
+			localEvList []types.Evidence
+			size        int64
 		)
 
-		pool, err := evidence.NewPool(logger, evidenceDB, stateStores[i], blockStore)
-		require.NoError(t, err)
+		// wait till we have at least the amount of evidence
+		// that we expect. if there's more local evidence then
+		// it doesn't make sense to wait longer and a
+		// different assertion should catch the resulting error
+		loops := 0
+		for len(localEvList) < len(evList) {
+			// each evidence should not be more than 500 bytes
+			localEvList, size = pool.PendingEvidence(int64(len(evList) * 500))
+			if loops == 10 {
+				t.Log("current wait status:", "|",
+					"local", len(localEvList), "|",
+					"waitlist", len(evList), "|",
+					"size", size)
+			}
+			if loops == 10000 && size == 0 {
+				assert.Fail(t, "pool seems disconnected")
+				return
+			}
 
-		testSuites[i] = setup(t, logger, pool, chBuf)
-	}
+			loops++
+		}
 
-	return testSuites
-}
+		// put the reaped evidence in a map so we can quickly check we got everything
+		evMap := make(map[string]types.Evidence)
+		for _, e := range localEvList {
+			evMap[string(e.Hash())] = e
+		}
 
-func waitForEvidence(t *testing.T, evList types.EvidenceList, suites ...*reactorTestSuite) {
-	t.Helper()
+		for i, expectedEv := range evList {
+			gotEv := evMap[string(expectedEv.Hash())]
+			require.Equalf(
+				t,
+				expectedEv,
+				gotEv,
+				"evidence for pool %d in pool does not match; got: %v, expected: %v", i, gotEv, expectedEv,
+			)
+		}
+	}
 
-	wg := new(sync.WaitGroup)
+	if len(ids) == 1 {
+		// special case waiting once, just to avoid the extra
+		// goroutine, in the case that this hits a timeout,
+		// the stack will be clearer.
+		fn(rts.pools[ids[0]])
+		return
+	}
 
-	for _, suite := range suites {
-		wg.Add(1)
+	wg := sync.WaitGroup{}
 
-		go func(s *reactorTestSuite) {
-			var localEvList []types.Evidence
+	for id := range rts.pools {
+		if len(ids) > 0 && !p2ptest.NodeInSlice(id, ids) {
```

Comment:
When would the ID not be in the slice?

Reply:
The slice is the input to the wait function, and the legacy implementation of this function let you wait for evidence to reach a specific set of nodes. So this if says "if no nodes are specified wait for the evidence to reach all nodes, otherwise, skip nodes in the network that aren't in the list specified. 

I've added a comment to make this more clear.

Reply:
Got it 👍